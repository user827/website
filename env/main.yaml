name: website
ssm_config_version: 2
