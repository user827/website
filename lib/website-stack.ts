import * as cdk from 'aws-cdk-lib';

import {
  aws_route53_targets as targets,
  aws_certificatemanager as acm,
  aws_cloudwatch as cloudwatch,
  aws_cloudwatch_actions as actions,
  aws_sns as sns,
  aws_sns_subscriptions as subscriptions,
  aws_s3 as s3,
  aws_s3_deployment as s3deploy,
  aws_route53 as route53,
  aws_iam as iam,
  aws_cloudfront as cloudfront,
  aws_cloudfront_origins as cloudfront_origins,
  RemovalPolicy,
  Duration,
  CfnOutput,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface WebsiteProps extends cdk.StackProps {
  cloudfrontOAI: cloudfront.IOriginAccessIdentity;
  siteBucket: s3.IBucket;
  certificate: acm.ICertificate;
  zone: string;
  email: string;
}

export class WebsiteStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: WebsiteProps) {
    super(scope, id, props);

    const zone = route53.HostedZone.fromLookup(this, 'Zone', { domainName: props.zone });

    const logBucket = new s3.Bucket(this, 'LogBucket', {
      encryption: s3.BucketEncryption.S3_MANAGED,
      enforceSSL: true,
      publicReadAccess: false,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      removalPolicy: RemovalPolicy.RETAIN,
    });
    logBucket.addLifecycleRule({
      expiration: Duration.days(7),
    });

    // CloudFront distribution
    const distribution = new cloudfront.Distribution(this, 'SiteDistribution', {
      certificate: props.certificate,
      logBucket,
      defaultRootObject: 'index.html',
      domainNames: [zone.zoneName, 'www.'+zone.zoneName],
      errorResponses: [
        {
          httpStatus: 403,
          responseHttpStatus: 404,
          responsePagePath: '/error.html',
          ttl: Duration.minutes(30),
        },
      ],
      defaultBehavior: {
        origin: new cloudfront_origins.S3Origin(props.siteBucket, { originAccessIdentity: props.cloudfrontOAI }),
        compress: true,
        allowedMethods: cloudfront.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
        viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
      },
    });

    new CfnOutput(this, 'DistributionId', { value: distribution.distributionId });

    // Route53 alias record for the CloudFront distribution
    new route53.ARecord(this, 'SiteApexAliasRecord', {
      target: route53.RecordTarget.fromAlias(new targets.CloudFrontTarget(distribution)),
      zone,
    });

    new route53.ARecord(this, 'SiteAliasRecord', {
      recordName: 'www',
      target: route53.RecordTarget.fromAlias(new targets.CloudFrontTarget(distribution)),
      zone,
    });

    // Deploy site contents to S3 bucket
    new s3deploy.BucketDeployment(this, 'DeployWithInvalidation', {
      sources: [s3deploy.Source.asset('./site-contents')],
      destinationBucket: props.siteBucket,
      distribution,
      distributionPaths: ['/*'],
    });


    this.createDashboard(distribution, props.email);
  }

  createDashboard(distribution: cloudfront.IDistribution, email: string): void {
    const t = new sns.Topic(this, 'TopicRule', {});

    t.addSubscription(new subscriptions.EmailSubscription(email));

    // Dashboard configuration
    const dashboard = new cloudwatch.Dashboard(this, 'Dashboard', {
      dashboardName: this.node.path.replace(/\//g, '-'),
    });

    const mRequests = new cloudwatch.Metric({
      namespace: 'AWS/CloudFront',
      metricName: 'Requests',
      period: cdk.Duration.minutes(1),
      statistic: cloudwatch.Stats.SUM,
      dimensionsMap: {
        DistributionId: distribution.distributionId,
        Region: 'Global',
      },
      region: 'us-east-1',
    });
    const aRequests = mRequests.createAlarm(this, 'AlarmRequests', {
      threshold: 1000,
      evaluationPeriods: 1,
    });
    aRequests.addAlarmAction(new actions.SnsAction(t));

    const wRequests = new cloudwatch.GraphWidget({
      title: 'Requests',
      left: [
        mRequests,
      ],
    });

    const m4xxErrors = new cloudwatch.Metric({
      namespace: 'AWS/CloudFront',
      metricName: '4xxErrorRate',
      period: cdk.Duration.minutes(1),
      statistic: cloudwatch.Stats.AVERAGE,
      dimensionsMap: {
        DistributionId: distribution.distributionId,
        Region: 'Global',
      },
      region: 'us-east-1',
    });

    const m5xxErrors = new cloudwatch.Metric({
      namespace: 'AWS/CloudFront',
      metricName: '5xxErrorRate',
      period: cdk.Duration.minutes(1),
      statistic: cloudwatch.Stats.AVERAGE,
      dimensionsMap: {
        DistributionId: distribution.distributionId,
        Region: 'Global',
      },
      region: 'us-east-1',
    });

    const wErrors = new cloudwatch.GraphWidget({
      stacked: true,
      title: 'Errors',
      left: [
        m4xxErrors,
      ],
      right: [
        m5xxErrors,
      ],
    });

    dashboard.addWidgets(wRequests, wErrors);
  }
}

export interface CertificateProps extends cdk.StackProps {
  zone: string;
  existingCertificateARN: string | undefined;
}


export class CertificateStack extends cdk.Stack {
  public readonly certificate: acm.ICertificate;

  constructor(scope: Construct, id: string, props: CertificateProps) {
    super(scope, id, props);
    if (!props.env) {
      throw new Error('need region');
    }
    if (props.env.region != 'us-east-1') {
      throw new Error('wrong region');
    }

    const zone = route53.HostedZone.fromLookup(this, 'Zone', { domainName: props.zone });

    // Certificates are expensive to make so try to do it only once and for all
    if (props.existingCertificateARN) {
      this.certificate = acm.Certificate.fromCertificateArn(this, 'SiteCertificate', props.existingCertificateARN);
    } else {
      this.certificate = new acm.Certificate(this, 'SiteCertificate', {
        domainName: zone.zoneName,
        subjectAlternativeNames: ['www.'+zone.zoneName],
        validation: acm.CertificateValidation.fromDns(zone),
      });
      this.certificate.applyRemovalPolicy(RemovalPolicy.RETAIN);
    }
    new CfnOutput(this, 'Certificate', {
      value: this.certificate.certificateArn,
    });
  }
}

export interface S3Props extends cdk.StackProps {
  zone: string;
}

export class S3Stack extends cdk.Stack {
  public readonly siteBucket: s3.IBucket;
  public readonly cloudfrontOAI: cloudfront.OriginAccessIdentity;

  constructor(scope: Construct, id: string, props: S3Props) {
    super(scope, id, props);

    this.cloudfrontOAI = new cloudfront.OriginAccessIdentity(this, 'CloudfrontOAI', {
      comment: `OAI for ${id}`,
    });

    this.siteBucket = new s3.Bucket(this, 'SiteBucket', {
      encryption: s3.BucketEncryption.S3_MANAGED,
      enforceSSL: true,
      publicReadAccess: false,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });
    new CfnOutput(this, 'Bucket', { value: this.siteBucket.bucketName });

    // Grant access to cloudfront
    const result = this.siteBucket.addToResourcePolicy(new iam.PolicyStatement({
      actions: ['s3:GetObject'],
      resources: [this.siteBucket.arnForObjects('*')],
      principals: [new iam.CanonicalUserPrincipal(this.cloudfrontOAI.cloudFrontOriginAccessIdentityS3CanonicalUserId)],
    }));

    if (!result) {
      throw new Error('addToResourcePolicy failed');
    }
  }
}

// TODO use arn sourcearn for OAC instead of the deprecated OAI
// https://github.com/aws/aws-cdk/issues/21771
// https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-restricting-access-to-s3.html

// https://aws.amazon.com/premiumsupport/knowledge-center/cloudfront-serve-static-website/
