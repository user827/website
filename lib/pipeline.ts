import * as cdk from 'aws-cdk-lib';
import {
  aws_iam as iam,
  aws_ssm as ssm,
  pipelines,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { WebsiteStage } from './website-stage';

export interface PipelineStackProps extends cdk.StackProps {
  zone: string;
  name: string;
  repo: string;
  branch: string;
  email: string | undefined;
}

export class PipelineStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: PipelineStackProps) {
    super(scope, id, props);

    // For the lookups to work
    if (!props.env) {
      throw new Error('env needs to be defined');
    }

    const pipeline = new pipelines.CodePipeline(this, 'Pipeline', {
      pipelineName: id,
      synth: new pipelines.CodeBuildStep('Synth', {
        installCommands: ['npm i -g npm@latest'],
        input: pipelines.CodePipelineSource.connection(props.repo, props.branch, {
          connectionArn: ssm.StringParameter.valueForStringParameter(this, `/${props.name}/connectionArn`),
          codeBuildCloneOutput: true,
        }),
        env: {
          PUBKEY: ssm.StringParameter.valueForStringParameter(this, `/${props.name}/pubkey`),
        },
        commands: [
          'echo "$PUBKEY" | gpg --import',
          'git verify-commit -v HEAD',
          'npm ci',
          'npm run lint',
          'npx cdk synth',
        ],
        rolePolicyStatements: [
          new iam.PolicyStatement({
            actions: ['sts:AssumeRole'],
            resources: ['*'],
            conditions: {
              StringEquals: {
                'iam:ResourceTag/aws-cdk:bootstrap-role': 'lookup',
              },
            },
          }),
        ],
      }),
    });

    if (!props.email) {
      props.email = ssm.StringParameter.valueFromLookup(this, `/${props.name}/email`);
    }
    const website = new WebsiteStage(this, 'Website', {
      env: props.env,
      zone: props.zone,
      email: props.email,
    });
    cdk.Tags.of(website).add('Project', props.name);
    pipeline.addStage(website);
  }
}
