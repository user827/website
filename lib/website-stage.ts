import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { WebsiteStack, S3Stack } from '../lib/website-stack';

export interface WebsiteProps extends cdk.StageProps {
  zone: string;
  hostedZoneId: string;
  email: string;
}

export class WebsiteStage extends cdk.Stage {
  public readonly s3: S3Stack;
  public readonly website: WebsiteStack;

  constructor(scope: Construct, id: string, props: WebsiteProps) {
    super(scope, id, props);

    if (!props.env) {
      throw new Error('env needs to be defined');
    }
    const env = props.env;

    this.s3 = new S3Stack(this, 'S3Stack', {
      env: { account: env.account, region: 'us-east-1' },
      zone: props.zone,
    });

    this.website = new WebsiteStack(this, 'WebsiteStack', {
      env: { account: env.account, region: 'us-east-1' },
      cloudfrontOAI: this.s3.cloudfrontOAI,
      siteBucket: this.s3.siteBucket,
      zone: props.zone,
      hostedZoneId: props.hostedZoneId,
      email: props.email,
    });
  }
}

export interface CertificateProps extends cdk.StageProps {
  zone: string;
  existingCertificateARN: string | undefined;
}
