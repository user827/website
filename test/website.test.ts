import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { testEnv } from './env';
import { WebsiteStage } from '../lib/website-stage';

// example test. To run these tests, uncomment this file along with the
// example resource in lib/website-stack.ts

const app = new cdk.App();
const stage = new WebsiteStage(app, 'test', testEnv);

test('s3', () => {
  const template = Template.fromStack(stage.s3);
  expect(template.toJSON()).toMatchSnapshot();

//   template.hasResourceProperties('AWS::SQS::Queue', {
//     VisibilityTimeout: 300
//   });
});

test('website', () => {
  const template = Template.fromStack(stage.website);
  expect(template.toJSON()).toMatchSnapshot();

//   template.hasResourceProperties('AWS::SQS::Queue', {
//     VisibilityTimeout: 300
//   });
});
