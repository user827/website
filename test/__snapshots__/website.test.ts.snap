// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`s3 1`] = `
{
  "Outputs": {
    "Bucket": {
      "Value": {
        "Ref": "SiteBucket397A1860",
      },
    },
    "ExportsOutputFnGetAttSiteBucket397A1860ArnB404F589": {
      "Export": {
        "Name": "test-S3Stack:ExportsOutputFnGetAttSiteBucket397A1860ArnB404F589",
      },
      "Value": {
        "Fn::GetAtt": [
          "SiteBucket397A1860",
          "Arn",
        ],
      },
    },
    "ExportsOutputFnGetAttSiteBucket397A1860RegionalDomainName3CF658E3": {
      "Export": {
        "Name": "test-S3Stack:ExportsOutputFnGetAttSiteBucket397A1860RegionalDomainName3CF658E3",
      },
      "Value": {
        "Fn::GetAtt": [
          "SiteBucket397A1860",
          "RegionalDomainName",
        ],
      },
    },
    "ExportsOutputRefCloudfrontOAI6D521D0D62AFE01C": {
      "Export": {
        "Name": "test-S3Stack:ExportsOutputRefCloudfrontOAI6D521D0D62AFE01C",
      },
      "Value": {
        "Ref": "CloudfrontOAI6D521D0D",
      },
    },
    "ExportsOutputRefSiteBucket397A1860ADBF1315": {
      "Export": {
        "Name": "test-S3Stack:ExportsOutputRefSiteBucket397A1860ADBF1315",
      },
      "Value": {
        "Ref": "SiteBucket397A1860",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CloudfrontOAI6D521D0D": {
      "Properties": {
        "CloudFrontOriginAccessIdentityConfig": {
          "Comment": "OAI for S3Stack",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "SiteBucket397A1860": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "aws-cdk:cr-owned:5e63352d",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "SiteBucketPolicy3AC1D0F8": {
      "Properties": {
        "Bucket": {
          "Ref": "SiteBucket397A1860",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "SiteBucket397A1860",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "SiteBucket397A1860",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "CloudfrontOAI6D521D0D",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "SiteBucket397A1860",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`website 1`] = `
{
  "Mappings": {
    "AWSCloudFrontPartitionHostedZoneIdMap": {
      "aws": {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Outputs": {
    "DistributionId": {
      "Value": {
        "Ref": "SiteDistribution3FF9535D",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Alarm5xxErrors5DABB075": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "TopicRule40A4EA44",
          },
        ],
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 2,
        "Dimensions": [
          {
            "Name": "DistributionId",
            "Value": {
              "Ref": "SiteDistribution3FF9535D",
            },
          },
          {
            "Name": "Region",
            "Value": "Global",
          },
        ],
        "EvaluationPeriods": 5,
        "MetricName": "5xxErrorRate",
        "Namespace": "AWS/CloudFront",
        "Period": 600,
        "Statistic": "Average",
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "AlarmRequestsC46A35F4": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "TopicRule40A4EA44",
          },
        ],
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 2,
        "Dimensions": [
          {
            "Name": "DistributionId",
            "Value": {
              "Ref": "SiteDistribution3FF9535D",
            },
          },
          {
            "Name": "Region",
            "Value": "Global",
          },
        ],
        "EvaluationPeriods": 6,
        "MetricName": "Requests",
        "Namespace": "AWS/CloudFront",
        "Period": 600,
        "Statistic": "Sum",
        "Threshold": 100,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": {
      "DependsOn": [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-12345-us-east-1",
          "S3Key": "9eb41a5505d37607ac419321497a4f8c21cf0ee1f9b4a6b29aa04301aea5c7fd.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_CA_BUNDLE": "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "DeployWithInvalidationAwsCliLayerDEDD5787",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::cdk-hnb659fds-assets-12345-us-east-1",
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::cdk-hnb659fds-assets-12345-us-east-1/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::ImportValue": "test-S3Stack:ExportsOutputFnGetAttSiteBucket397A1860ArnB404F589",
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::ImportValue": "test-S3Stack:ExportsOutputFnGetAttSiteBucket397A1860ArnB404F589",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "cloudfront:GetInvalidation",
                "cloudfront:CreateInvalidation",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": [
          {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "Dashboard9E4231ED": {
      "Properties": {
        "DashboardBody": {
          "Fn::Join": [
            "",
            [
              "{"widgets":[{"type":"metric","width":6,"height":6,"x":0,"y":0,"properties":{"view":"timeSeries","title":"Requests","region":"",
              {
                "Ref": "AWS::Region",
              },
              "","metrics":[["AWS/CloudFront","Requests","DistributionId","",
              {
                "Ref": "SiteDistribution3FF9535D",
              },
              "","Region","Global",{"period":600,"stat":"Sum"}]],"yAxis":{}}},{"type":"metric","width":6,"height":6,"x":6,"y":0,"properties":{"view":"timeSeries","title":"Errors","region":"",
              {
                "Ref": "AWS::Region",
              },
              "","stacked":true,"metrics":[["AWS/CloudFront","4xxErrorRate","DistributionId","",
              {
                "Ref": "SiteDistribution3FF9535D",
              },
              "","Region","Global",{"period":60}],["AWS/CloudFront","5xxErrorRate","DistributionId","",
              {
                "Ref": "SiteDistribution3FF9535D",
              },
              "","Region","Global",{"period":600}]],"yAxis":{}}}]}",
            ],
          ],
        },
        "DashboardName": "test-WebsiteStack",
      },
      "Type": "AWS::CloudWatch::Dashboard",
    },
    "DeployWithInvalidationAwsCliLayerDEDD5787": {
      "Properties": {
        "Content": {
          "S3Bucket": "cdk-hnb659fds-assets-12345-us-east-1",
          "S3Key": "e2277687077a2abf9ae1af1cc9565e6715e2ebb62f79ec53aa75a1af9298f642.zip",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "DeployWithInvalidationCustomResourceE3FF7455": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DestinationBucketName": {
          "Fn::ImportValue": "test-S3Stack:ExportsOutputRefSiteBucket397A1860ADBF1315",
        },
        "DistributionId": {
          "Ref": "SiteDistribution3FF9535D",
        },
        "DistributionPaths": [
          "/*",
        ],
        "Prune": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": [
          "cdk-hnb659fds-assets-12345-us-east-1",
        ],
        "SourceObjectKeys": [
          "0bcfa98d8a0e93cb713bc2560116c890a0d46888981db0552e621bada72be8ae.zip",
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "LogBucketCC3B17E8": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LifecycleConfiguration": {
          "Rules": [
            {
              "ExpirationInDays": 7,
              "Status": "Enabled",
            },
          ],
        },
        "OwnershipControls": {
          "Rules": [
            {
              "ObjectOwnership": "ObjectWriter",
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "LogBucketPolicy900DBE48": {
      "Properties": {
        "Bucket": {
          "Ref": "LogBucketCC3B17E8",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "LogBucketCC3B17E8",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "LogBucketCC3B17E8",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "SiteAliasRecord3C0AF5BF": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "SiteDistribution3FF9535D",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "www.test.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "SiteApexAliasRecord7A87C51C": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "SiteDistribution3FF9535D",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "test.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "SiteCertificate38C247F6": {
      "Properties": {
        "DomainName": "test",
        "DomainValidationOptions": [
          {
            "DomainName": "test",
            "HostedZoneId": "DUMMY",
          },
          {
            "DomainName": "www.test",
            "HostedZoneId": "DUMMY",
          },
        ],
        "SubjectAlternativeNames": [
          "www.test",
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/WebsiteStack/SiteCertificate",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "SiteDistribution3FF9535D": {
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "test",
            "www.test",
          ],
          "CustomErrorResponses": [
            {
              "ErrorCachingMinTTL": 1800,
              "ErrorCode": 403,
              "ResponseCode": 404,
              "ResponsePagePath": "/error.html",
            },
          ],
          "DefaultCacheBehavior": {
            "AllowedMethods": [
              "GET",
              "HEAD",
              "OPTIONS",
            ],
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "TargetOriginId": "testWebsiteStackSiteDistributionOrigin15B6C2BD0",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Logging": {
            "Bucket": {
              "Fn::GetAtt": [
                "LogBucketCC3B17E8",
                "RegionalDomainName",
              ],
            },
          },
          "Origins": [
            {
              "DomainName": {
                "Fn::ImportValue": "test-S3Stack:ExportsOutputFnGetAttSiteBucket397A1860RegionalDomainName3CF658E3",
              },
              "Id": "testWebsiteStackSiteDistributionOrigin15B6C2BD0",
              "S3OriginConfig": {
                "OriginAccessIdentity": {
                  "Fn::Join": [
                    "",
                    [
                      "origin-access-identity/cloudfront/",
                      {
                        "Fn::ImportValue": "test-S3Stack:ExportsOutputRefCloudfrontOAI6D521D0D62AFE01C",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "ViewerCertificate": {
            "AcmCertificateArn": {
              "Ref": "SiteCertificate38C247F6",
            },
            "MinimumProtocolVersion": "TLSv1.2_2021",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "TopicRule40A4EA44": {
      "Type": "AWS::SNS::Topic",
    },
    "TopicRuletesttesttest4EBD90B9": {
      "Properties": {
        "Endpoint": "test@test.test",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "TopicRule40A4EA44",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
